src/read.c

```diff
@@ -4342,41 +4351,49 @@ static avifResult avifDecoderDataFindToneMappedImageItem(const avifDecoderData *
     return AVIF_RESULT_OK;
 }
 
-// Finds a 'tmap' (tone mapped image item) box associated with the given 'colorItem',
+// Finds a 'tmap' (tone mapped image item) box associated with the given colorItem at index 'colorItemIndex',
 // then finds the associated gain map image.
-// If found, fills 'toneMappedImageItem', 'gainMapItem' and  'gainMapCodecType'.
-// Otherwise, sets 'toneMappedImageItem' and 'gainMapItem' to NULL.
+// If found, fills 'toneMappedImageItem', 'gainMapItemIndex' and  'gainMapCodecType'.
+// Otherwise, sets 'toneMappedImageItem' to NULL and 'gainMapItemIndex' to -1.
 // Returns AVIF_RESULT_OK if no errors were encountered (whether or not a gain map was found).
 // Assumes that there is a single tmap item, and not, e.g., a grid of tmap items.
 static avifResult avifDecoderFindGainMapItem(const avifDecoder * decoder,
-                                             const avifDecoderItem * colorItem,
+                                             int colorItemIndex,
                                              avifDecoderItem ** toneMappedImageItem,
-                                             avifDecoderItem ** gainMapItem,
+                                             int * gainMapItemIndex,
                                              avifCodecType * gainMapCodecType)
 {
     *toneMappedImageItem = NULL;
-    *gainMapItem = NULL;
+    *gainMapItemIndex = -1;
     *gainMapCodecType = AVIF_CODEC_TYPE_UNKNOWN;
 
     avifDecoderData * data = decoder->data;
 
     uint32_t gainMapItemID;
     avifDecoderItem * toneMappedImageItemTmp;
-    AVIF_CHECKRES(avifDecoderDataFindToneMappedImageItem(data, colorItem, &toneMappedImageItemTmp, &gainMapItemID));
+    AVIF_CHECKRES(
+        avifDecoderDataFindToneMappedImageItem(data, &data->meta->items.item[colorItemIndex], &toneMappedImageItemTmp, &gainMapItemID));
     if (!toneMappedImageItemTmp) {
         return AVIF_RESULT_OK;
     }
 
     assert(gainMapItemID != 0);
-    avifDecoderItem * gainMapItemTmp;
-    AVIF_CHECKRES(avifMetaFindOrCreateItem(data->meta, gainMapItemID, &gainMapItemTmp));
-    if (avifDecoderItemShouldBeSkipped(gainMapItemTmp)) {
+    avifDecoderItem * gainMapItem;
+    AVIF_CHECKRES(avifMetaFindOrCreateItem(data->meta, gainMapItemID, &gainMapItem));
+    if (avifDecoderItemShouldBeSkipped(gainMapItem)) {
         avifDiagnosticsPrintf(data->diag, "Box[tmap] gain map item %d is not a supported image type", gainMapItemID);
         return AVIF_RESULT_INVALID_TONE_MAPPED_IMAGE;
     }
+    for (uint32_t itemIndex = 0; itemIndex < data->meta->items.count; ++itemIndex) {
+        if (gainMapItem->id == data->meta->items.item[itemIndex].id) {
+            *gainMapItemIndex = itemIndex;
+            break;
+        }
+    }
+    assert(*gainMapItemIndex >= 0);
```

Comment:
Nit/optional: It is unfortunate that we have to search in the `data->meta->items` array to find the item index. We can consider adding a variant of the `avifMetaFindOrCreateItem` function that outputs an item index rather than an item pointer.

Reply:
Ack, i thought about this as well. But since this is the only use, i will leave this as-is.